### PR TITLE
Filters v23

### PR DESCRIPTION
**Pull request by jamesRUS52**
Fix and Improve filters

### CODE REVIEW

**Review by luanfreitasdev — CHANGES_REQUESTED** [review]
As the tests are failing, maybe you need to review some things in the filters, you could run the composer check command to run them all or debug the tests that are failing using the ->only() method at the end of the test itself

https://pestphp.com/docs/skipping-tests#running-a-single-test

Datetime filter:

`if (isset($data['enableTime']) && !$data['enableTime'])`  ?

https://flareapp.io/share/A7EldMgP#F57

When changing pages:
image

Tailwind problem because of Jquery:
image

L18: https://github.com/Power-Components/livewire-powergrid/pull/294/commits/03a6f7e238d11190c623e14292bf83d5558a1a61#diff-9facb2e443abc6b4fbb6b782046b8b5f1c6adfeee68442809e5a00903e8e075a

```php
@if(isBootstrap5())
 script for bootstrap
@endif
```

**Comment by jamesRUS52** [comment]
This two error occure because new js file was missing in commit.
added it

**Comment by jamesRUS52** [comment]
I don't know why tests are fails
I did't change any in text filter

**Comment by luanfreitasdev** [comment]
> This two error occure because new js file was missing in commit. added it

script.blade it's global. you must specify which theme will use jquery code

**Comment by luanfreitasdev** [comment]
The problem is related to asserting the filters variable which is no longer empty:

FilterInputOptionsModelTest.php
`->assertViewHas('filters', []); // It's not empty anymore`

Because the filter is not empty, we have some problems shown in the video

I believe the current logic is better in the sense that when we click on clear a column, the text or selection should also be cleared:

current:
https://user-images.githubusercontent.com/33601626/154804393-408033e2-1f9d-43a9-9d5a-9c1e874f1aba.mp4

now:
https://user-images.githubusercontent.com/33601626/154804339-f978473b-bbe9-46be-a7ed-b8a2e62f7a77.mp4

We need to fix this issue and replicate the same logic for tailwind

**Comment by jamesRUS52** [comment]
added only for bootstrap

**Comment by jamesRUS52** [comment]
> now:
> https://user-images.githubusercontent.com/33601626/154804339-f978473b-bbe9-46be-a7ed-b8a2e62f7a77.mp4

Is it after last commit?
I can't reproduce it in both bootstrap and tailwind themes

**Comment by luanfreitasdev** [comment]
> > now:
> > https://user-images.githubusercontent.com/33601626/154804339-f978473b-bbe9-46be-a7ed-b8a2e62f7a77.mp4
> 
> Is it after last commit? I can't reproduce it in both bootstrap and tailwind themes

Yeah, that one was just to not break the tailwind.

**Comment by jamesRUS52** [comment]
I have no idea why test are failing.
Are the tests correct? It look like test was success earlier because
```
$this->filters = [];
```
clear all filters always by click by any filter in list

**Comment by luanfreitasdev** [comment]
I downloaded the branch and the problem is just cleaning up the `->makeInputText()` all the others are right.

however we have the problem of `if (!$data['enableTime'])`

And replicate the logic for tailwind

**Comment by jamesRUS52** [comment]
```
if (!$data['enableTime'])
```
It not for bootstrap only.
this param must be in both theme because pg-flat-pickr.js common.
the problem must be gone after last commit where I done 
```
npm run build
```
for changed pg-flat-pickr.js

**Comment by jamesRUS52** [comment]
> the problem is just cleaning up the `->makeInputText()`

sorry, I don't understand what do you mean.
Can you show the code, where it should be fixed ?

**Comment by luanfreitasdev** [comment]
When you clear the text field (generated by makeInputText) the content should also be cleared.

As much as it is applied to bootstrap, the backend changes have changed the behavior for both

**Comment by jamesRUS52** [comment]
I can't repeat it.
In my demo stand all work well in bootstrap and tailwind.
When I clear the input field it clear filter and input value

**Comment by jamesRUS52** [comment]
I have fixed the test

**Comment by jamesRUS52** [comment]
[video](https://vimeo.com/679537042)

**Comment by jamesRUS52** [comment]
Do you wait anything from me?

**Comment by luanfreitasdev** [comment]
Thanks, but I need to test again

**Comment by jamesRUS52** [comment]
I resolved problem with
```if (!$data['enableTime'])
```
if we have locale != en for datetime picker and it not loaded
or
if this locale does not have enableTime option
the error occure

I added default value 'false'  to submit event method

File: src/Traits/Filter.php [hunk 1]
```diff
@@ -19,17 +19,25 @@ trait Filter
...
+        unset($this->filters['number_end'][$field]);
         unset($this->filters['input_text'][$field]);
         unset($this->filters['boolean'][$field]);
         unset($this->filters['input_text_options'][$field]);
-
-        $this->filters = [];
+        unset($this->filters['input_date_picker'][$field]);
+        unset($this->filters['date_picker'][$field]);
+        unset($this->filters['select'][$field]);
+        unset($this->filters['multi_select'][$field]);
```
Comment (luanfreitasdev) [inline src/Traits/Filter.php:32]: I tested your branch, only the filter variable is blocking to follow. It should be like before, do you agree?

```
unset($this->enabledFilters[$field]);
$this->filters = [];
```

**Review by luanfreitasdev — CHANGES_REQUESTED** [review]

File: src/Traits/Filter.php (hunk 1 above)
Comment (jamesRUS52) [reply src/Traits/Filter.php:32]: No, it main bug. It remove all filters instead of remove one selected

Comment (luanfreitasdev) [reply src/Traits/Filter.php:32]: For me it is not working, if you do the same test in the browser that is in the pest test, it will give the same error

Comment (luanfreitasdev) [reply src/Traits/Filter.php:32]: The problem happens with tables that use joins with fields of this type: `dishes.name`. The proposed unset does not clean. Could you clean the first part of the array? 

unset($this->filters['input_text'][$field])

**Comment by luanfreitasdev** [comment]
I believe it works now. I will test again

**Comment by jamesRUS52** [comment]
Thanks for merging all my PRs )

**Comment by luanfreitasdev** [comment]
Great work, thank you very much and more contributions will be welcome

**Comment by jamesRUS52** [comment]
Now, #298 works as expected?

File: resources/views/components/frameworks/bootstrap5/table-base.blade.php [hunk 2]
```diff
@@ -5,15 +5,14 @@
...
-        <div class="table-responsive col-md-12" style="margin: 10px 0 10px;">
-            @include($table)
-        </div>
-        <div class="row">
-            <div class="col-12 overflow-auto">
-                @include($theme->footer->view, ['theme' => $theme])
-            </div>
+        @includeIf($theme->layout->message)
+    </div>
+    <div class="table-responsive overflow-visible col-md-12" style="margin: 10px 0 10px;">
```
Comment (luanfreitasdev) [inline resources/views/components/frameworks/bootstrap5/table-base.blade.php:10]: removed overflow-visible because it broke bootstrap layout on mobile

**Comment by luanfreitasdev** [comment]
> Now, #298 works as expected?

I'll test it again, but I believe it worked. I will make a release with this correction and the next one will be the persist data

File: resources/views/components/frameworks/bootstrap5/table-base.blade.php (hunk 2 above)
Comment (jamesRUS52) [reply resources/views/components/frameworks/bootstrap5/table-base.blade.php:10]: Could you suggest any alternatives for this?
The problem is that if we don't have rows in the table (or there are very few of them), the height of the table is too small to display a long list of the select/multiselect/datepicker filter well

**Pull request merged**